File: Ywy/Go/level1/제일 작은 수 제거하기/MinNumberRemove.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(solution([]int{4, 3, 2, 1}))
	fmt.Println(solution([]int{10}))

	fmt.Println(solution2([]int{4, 3, 2, 1}))
	fmt.Println(solution2([]int{10}))
}

func solution(arr []int) []int {
	minValueEl := arr[0]
	removeIndex := 0
	for index := range arr {

		if arr[index] < minValueEl {
			minValueEl = arr[index]
			removeIndex = index
		}
	}

	answer := append(arr[:removeIndex], arr[removeIndex+1:]...)

	if len(answer) == 0 {
		return []int{-1}
	}
	return answer
}

func solution2(arr []int) []int {
	size := len(arr)

	if size == 1 {
		return []int{-1}
	}

	minValueEl := arr[0]
	el := 0
	index := 0
	removeIndex := 0

	for {
		if index == size {
			return append(arr[:removeIndex], arr[removeIndex+1:]...)
		}

		el = arr[index]

		if el < minValueEl {
			minValueEl = el
			removeIndex = index
		}

		index++
	}
}

File: Ywy/JAVA/src/level1/MinNumberRemove.java
package level1;

public class MinNumberRemove {
	public int[] solution(final int[] arr) {
		int range = arr.length;
		if (range == 1) {
			return new int[] {-1};
		}
		
		int minValueIndex = getMinValue(arr, --range);
		final int[] answer = new int[range];
		int index = 0;
		int index2 = -1;
		while(true) {
			if (++index2 == minValueIndex) {
				continue;
			}
			answer[index++] = arr[index2];
			if (index == range) {
				return answer;
			}
		}
	}
	
	private int getMinValue(final int[] arr, final int range) {
		int index = 0;
		int minIndex = 0;
		int minValue = arr[index];
		while(true) {
			if (index == range) {
				return minIndex;
			}
			if (minValue > arr[++index]) {
				minValue = arr[index];
				minIndex = index;
			}
		}
	}
}
